SG14_test/inplace_function_test.cpp

```diff
@@ -489,6 +489,90 @@ static void test_return_by_move()
     assert(InstrumentedCopyConstructor::moves == 1);
 }
 
+struct NoDefaultCtor
+{
+    int val;
+    explicit NoDefaultCtor(int v) : val{v} {}
+};
+
+int overloaded_function(stdext::inplace_function<int()>&& exec)
+{
+    return exec();
+}
+
+int overloaded_function(stdext::inplace_function<int(int)>&& exec)
+{
+    return exec(42);
+}
+
+using C_Int1 = int();
+using C_Int2 = int(int);
+using C_Void = void(int&);
+
+using stdext::inplace_function_detail::is_invocable_r;
+
+static_assert(is_invocable_r<int, C_Int1>::value, "");
+static_assert(! is_invocable_r<int, C_Int2>::value, "");
+static_assert(! is_invocable_r<int, C_Void>::value, "");
+
+static_assert(is_invocable_r<int, C_Int2, int>::value, "");
+static_assert(! is_invocable_r<int, C_Int1, int>::value, "");
+static_assert(! is_invocable_r<int, C_Void, int>::value, "");
+
+static_assert(is_invocable_r<void, C_Void, int&>::value, "");
+static_assert(! is_invocable_r<void, C_Int1, int&>::value, "");
+
+// Testing widening and narrowing conversions, and the "conversion" to void.
+static_assert(is_invocable_r<void, C_Int1>::value, "");
+static_assert(is_invocable_r<long, C_Int1>::value, "");
+static_assert(is_invocable_r<char, C_Int1>::value, "");
+
+// Testing the conversion from void to int, which should definitely not be allowed.
+static_assert(! is_invocable_r<int, C_Void, int&>::value, "");
+
+// cppreference:
+// > Determines whether Fn can be invoked with the arguments ArgTypes...
+// > to yield a result that is convertible to R.
+//
+// void is treated specially because a functions return value can be ignored.
+static_assert(is_invocable_r<void, C_Int2, int&>::value, "");
+static_assert(is_invocable_r<const void, C_Int2, int&>::value, "");
+
+// Regression tests for both is_invocable and is_convertible.
+static_assert(is_invocable_r<const int&, int()>::value, "");
+static_assert(is_invocable_r<const int&, int(*)()>::value, "");
```

Comment:
I'd like to see lines 508–543 moved into a function scope, say, `test_is_invocable` (and then called from line 663-ish). That way the typedefs and the `using`-declaration won't leak out into the rest of the file.